File: php/OrdersSample.php
<?php
/**
 * Copyright 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

require_once 'BaseSample.php';

/**
 * Class for running through an example workflow with the
 * Orders service. For this sample, we use the sandbox API endpoint
 * so that we can create/manage test orders and so that we don't
 * accidentally mutate real orders in the system.
 */
class OrdersSample extends BaseSample {
  private $nonce = 0; // used by newOperationId()

  public function run() {
    // First, we create a new test order using the template1 template.
    // Normally, orders would be automatically populated by Google in the
    // non-sandbox version, and we'd skip ahead to find out what orders are
    // currently waiting for our acknowledgment.
    print 'Creating new test order...';
    $request =
        new Google_Service_ShoppingContent_OrdersCreateTestOrderRequest();
    $request->setTemplateName("template1");
    $resp = $this->sandboxService->orders->createtestorder(
        $this->merchantId, $request);
    $orderId = $resp->getOrderId();
    printf("done (%s).\n", $orderId);
    print "\n";

    // List all unacknowledged orders.  In normal usage, this is where we'd
    // get new order IDs to operate on. We should see the new test order
    // show up here.
    $this->listUnacknowledgedOrders();

    // Acknowledge the newly created order.  The assumption in doing so is that
    // we've collected the information for this order into our own internal
    // system, and acknowledging it allows us to skip it when searching for new
    // orders in the future.
    $this->acknowledge($orderId);

    // We'll also set the merchant order ID to simulate having entered it into
    // our own internal database.  For this, and the tracking and shipping IDs
    // we'll see later, we'll use random numbers.
    $this->updateMerchantOrderId($orderId, mt_rand());

    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";

    // Oops, we don't have enough stock for all the Chromecasts ordered! Let's
    // cancel one of them from the order.
    $this->cancelLineItem(
        $orderId,
        $currentOrder->getLineItems()[0]->getId(),
        1,
        'noInventory',
        'Ran out of inventory while fulfilling request.'
    );

    // Notice the change in quantities in this updated view of the order.
    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";

    // At this point, we'll advance the test order.  This simulates the point
    // at which an order is no longer cancellable by the customer and the order
    // information reflects that we're ready for the merchant to ship.
    printf("Advancing test order %s...", $orderId);
    $this->sandboxService->orders->advancetestorder(
        $this->merchantId, $orderId);
    print "done.\n";
    print "\n";

    // Notice the change in status for the order.
    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";

    // To simulate partial fulfillment, we'll pick the first line item and
    // ship the still-pending amount. (Remember, we cancelled one already.)
    // We store the request to later access the random strings generated
    // for the shipment and tracking IDs.
    $shipItem1Req = $this->shipLineItemAll(
        $orderId, $currentOrder->getLineItems()[0]);

    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";

    // Now we ship the rest.
    $shipItem2Req = $this->shipLineItemAll(
        $orderId, $currentOrder->getLineItems()[1]);

    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";

    // Customer receives the first item, so we notify Google that it was
    // delivered.
    $this->lineItemDelivered($orderId, $shipItem1Req);

    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";

    // Customer receives the second item, so we notify Google that it was
    // delivered.
    $this->lineItemDelivered($orderId, $shipItem2Req);

    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";

    // Turns out one of the first line item was damaged when it arrived,
    // so we let Google know the customer returned one.
    $this->lineItemReturned(
        $orderId,
        $currentOrder->getLineItems()[0]->getId(),
        1,
        'productArrivedDamaged',
        'Item was non-functional on receipt.'
    );

    $currentOrder = $this->getOrder($orderId);
    $this->printOrder($currentOrder);
    print "\n";
  }

  /**
   * Operation IDs (even across operations) must be unique over the lifetime
   * of an order, so that Google can detect and reject duplicate requests.
   * Since we're sending requests sequentially and not retrying, we just use
   * a simple nonce that's incremented each time.
   */
  private function newOperationId() {
    return strval($this->nonce++);
  }

  /**
   * Lists the unacknowledged orders for {@code $this->merchantId}, printing
   * out each in turn.
   */
  public function listUnacknowledgedOrders() {
    printf("Printing unacknowledged orders for %s.\n", $this->merchantId);
    $parameters = ['acknowledged' => 'false'];
    do {
      $resp = $this->sandboxService->orders->listOrders(
          $this->merchantId, $parameters);
      if (empty($resp->getResources())) {
        print ("No orders.\n");
      }
      foreach ($resp->getResources() as $order) {
        $this->printOrder($order);
      }
      $parameters['pageToken'] = $resp->getNextPageToken();
    } while (!empty($parameters['pageToken']));
    print("\n");
  }

  /**
   * Acknowledges order {@code $orderId}, which allows us to filter it out
   * when looking for new orders.
   *
   * @param string $orderId the order ID of the order to acknowledge
   */
  public function acknowledge($orderId) {
    printf("Acknowledging order %s... ", $orderId);
    $req = new Google_Service_ShoppingContent_OrdersAcknowledgeRequest();
    $req->setOperationId($this->newOperationId());
    $resp = $this->sandboxService->orders->acknowledge(
        $this->merchantId, $orderId, $req);
    printf("done (%s).\n", $resp->getExecutionStatus());
    print "\n";
  }

  /**
   * Retrieves the order information for {@code $orderId}.
   *
   * @param string $orderId the order ID of the order to retrieve
   * @return Google_Service_ShoppingContent_Order
   */
  public function getOrder($orderId) {
    printf("Retrieving order %s... ", $orderId);
    $order = $this->sandboxService->orders->get($this->merchantId, $orderId);
    print "done.\n";
    print "\n";
    return $order;
  }

  /**
   * Updates the merchant order ID for {@code $orderId}.
   *
   * @param string $orderId the order ID of the order to update
   * @param string $merchantOrderId the new merchant order ID of the order
   */
  public function updateMerchantOrderId($orderId, $merchantOrderId) {
    printf("Updating merchant order ID to %s... ", $merchantOrderId);
    $req =
        new Google_Service_ShoppingContent_OrdersUpdateMerchantOrderIdRequest();
    $req->setOperationId($this->newOperationId());
    $req->setMerchantOrderId($merchantOrderId);

    $resp = $this->sandboxService->orders->updatemerchantorderid(
        $this->merchantId, $orderId, $req);
    printf("done (%s).\n", $resp->getExecutionStatus());
    print "\n";
  }

  /**
   * Cancels a line item from the order {@code $orderId}.
   *
   * @param string $orderId the order ID of the order to update
   * @param string $lineItemId the ID of the line item to cancel
   * @param int $quantity amount of the line item to cancel
   * @param string $reason a value from a Google-defined enum (see docs)
   * @param string $reasonText free-form text explaining the cancellation
   *
   * @see https://developers.google.com/shopping-content/v2/reference/v2/orders/cancellineitem
   */
  public function cancelLineItem(
      $orderId, $lineItemId, $quantity, $reason, $reasonText) {
    printf("Cancelling %d of item %s... ", $quantity, $lineItemId);
    $req = new Google_Service_ShoppingContent_OrdersCancelLineItemRequest();
    $req->setLineItemId($lineItemId);
    $req->setQuantity($quantity);
    $req->setReason($reason);
    $req->setReasonText($reasonText);
    $req->operationId = $this->newOperationId();

    $resp = $this->sandboxService->orders->cancellineitem(
        $this->merchantId, $orderId, $req);
    printf("done (%s).\n", $resp->getExecutionStatus());
    print "\n";
  }

  /**
   * Marks a line item from the order {@code $orderId} as having shipped.
   * This method uses the pending quantity of the item.  It returns the
   * shipping request so we can access the randomly-generated tracking and
   * shipping IDs.
   *
   * @param string $orderId the order ID of the order to update
   * @param Google_Service_ShoppingContent_OrderLineItem $lineItem
   *        the line item to cancel
   * @return Google_Service_ShoppingContent_OrdersShipLineItemsRequest
   */
  public function shipLineItemAll($orderId, $lineItem) {
    printf("Shipping %d of item %s... ", $lineItem->getQuantityPending(),
        $lineItem->getId());
    $item = new Google_Service_ShoppingContent_OrderShipmentLineItemShipment();
    $item->setLineItemId($lineItem->getId());
    $item->setQuantity($lineItem->getQuantityPending());

    $req = new Google_Service_ShoppingContent_OrdersShipLineItemsRequest();
    $req->setCarrier(
        $lineItem->getShippingDetails()->getMethod()->getCarrier());
    $req->setShipmentId(mt_rand());
    $req->setTrackingId(mt_rand());
    $req->setLineItems([$item]);
    $req->operationId = $this->newOperationId();

    $resp = $this->sandboxService->orders->shiplineitems(
        $this->merchantId, $orderId, $req);
    printf("done (%s).\n", $resp->getExecutionStatus());
    print "\n";

    return $req;
  }

  /**
   * Marks all of the line item from the order {@code $orderId} that was
   * shipped as being delivered.
   *
   * @param string $orderId the order ID of the order to update
   * @param Google_Service_ShoppingContent_OrdersShipLineItemsRequest $shipReq
   *        the shipping information for the line item to update
   */
  public function lineItemDelivered($orderId, $shipReq) {
    printf("Marking shipment %s as delivered... ", $shipReq->getShipmentId());
    $req = new Google_Service_ShoppingContent_OrdersUpdateShipmentRequest();
    $req->setCarrier($shipReq->getCarrier());
    $req->setShipmentId($shipReq->getShipmentId());
    $req->setTrackingId($shipReq->getTrackingId());
    $req->setStatus('delivered');
    $req->operationId = $this->newOperationId();

    $resp = $this->sandboxService->orders->updateshipment(
        $this->merchantId, $orderId, $req);
    printf("done (%s).\n", $resp->getExecutionStatus());
    print "\n";
  }

  /**
   * Returns a line item from the order {@code $orderId}.
   *
   * @param string $orderId the order ID of the order to update
   * @param string $lineItemId the ID of the line item to return
   * @param int $quantity amount of the line item to return
   * @param string $reason a value from a Google-defined enum (see docs)
   * @param string $reasonText free-form text explaining the return
   *
   * @see https://developers.google.com/shopping-content/v2/reference/v2/orders/returnlineitem
   */
  public function lineItemReturned(
      $orderId, $lineItemId, $quantity, $reason, $reasonText) {
    printf("Marking %d of item %s as returned... ", $quantity, $lineItemId);
    $req = new Google_Service_ShoppingContent_OrdersReturnLineItemRequest();
    $req->setLineItemId($lineItemId);
    $req->setQuantity($quantity);
    $req->setReason($reason);
    $req->setReasonText($reasonText);
    $req->operationId = $this->newOperationId();

    $resp = $this->sandboxService->orders->returnlineitem(
        $this->merchantId, $orderId, $req);
    printf("done (%s).\n", $resp->getExecutionStatus());
    print "\n";
  }

  // Helper function to just make code more streamlined below.  Default for
  // $indent fits most uses.
  private function printIfTrue($value, $text, $indent = '  ') {
    if($value) {
      printf($indent . "- %s: %s\n", $text, $value);
    }
  }

  /**
   * Prints the order information contained in {@code $order}.
   *
   * @param Order $order the Order resource to print
   */
  public function printOrder($order) {
    printf("Order %s:\n", $order->getId());
    printf("- Status: %s\n", $order->getStatus());
    printf("- Merchant: %d\n", $order->getMerchantId());
    $this->printIfTrue($order->getMerchantOrderId(),
        'Merchant order ID', '');
    printf("- Placed on date: %s\n", $order->getPlacedDate());
    if ($order->getNetAmount()) {
      $amount = $order->getNetAmount();
      printf("- Net amount: %s %s\n", $amount->getValue(),
          $amount->getCurrency());
    }
    if ($order->getPaymentMethod()) {
      $method = $order->getPaymentMethod();
      print "- Payment method:\n";
      printf("  - Type: %s\n", $method->getType());
      printf("  - Expiration date: %s/%s\n", $method->getExpirationMonth(),
          $method->getExpirationYear());
    }
    printf("- Acknowledged: %s\n", $order->getAcknowledged() ? 'yes' : 'no');
    if (!empty($order->getLineItems())) {
      printf("- %d line item(s):\n", count($order->getLineItems()));
      foreach ($order->getLineItems() as $item) {
        $this->printLineItem($item);
      }
    }
    printf("- Shipping option: %s\n", $order->getShippingOption());
    if ($order->getShippingCost()) {
      $cost = $order->getShippingCost();
      printf("- Shipping cost: %s %s\n", $cost->getValue(),
          $cost->getCurrency());
    }
    if ($order->getShippingCostTax()) {
      $tax = $order->getShippingCostTax();
      printf("- Shipping cost: %s %s\n", $tax->getValue(), $tax->getCurrency());
    }
    if (!empty($order->getShipments())) {
      printf("- %d shipment(s):\n", count($order->getShipments()));
      foreach ($order->getShipments() as $shipment) {
        printf("  Shipment %s:\n", $shipment->getId());
        printf("  - Creation date: %s\n", $shipment->getCreationDate());
        printf("  - Carrier: %s\n", $shipment->getCarrier());
        printf("  - Tracking ID: %s\n", $shipment->getTrackingId());
        if (!empty($shipment->getLineItems())) {
          printf("  - %d line item(s):\n", count($shipment->getLineItems()));
          foreach ($shipment->getLineItems() as $item) {
            printf("    %d of item %s\n", $item->getQuantity(),
                $item->getLineItemId());
          }
        }
        $this->printIfTrue($shipment->getDeliveryDate(), "Delivery date");
      }
    }
  }

  private function printLineItem($item) {
    printf("  Line item: %s\n", $item->getId());
    printf("  - Product: %s (%s)\n", $item->getProduct()->getId(),
        $item->getProduct()->getTitle());
    printf("  - Price: %s %s\n", $item->getPrice()->getValue(),
        $item->getPrice()->getCurrency());
    printf("  - Tax: %s %s\n", $item->getTax()->getValue(),
        $item->getTax()->getCurrency());
    if ($item->getShippingDetails()) {
      $details = $item->getShippingDetails();
      printf("  - Ship by date: %s\n", $details->getShipByDate());
      printf("  - Deliver by date: %s\n", $details->getDeliverByDate());
      $method = $details->getMethod();
      printf("  - Deliver via %s %s (%s - %s days)\n",
          $method->getCarrier(), $method->getMethodName(),
          $method->getMinDaysInTransit(), $method->getMaxDaysInTransit());
    }
    if ($item->getReturnInfo() && $item->getReturnInfo()->getIsReturnable()) {
      $info = $item->getReturnInfo();
      printf("  - Item is returnable.\n");
      printf("    - Days to return: %s\n", $info->getDaysToReturn());
      printf("    - Return policy is at %s.\n", $info->getPolicyUrl());
    } else {
      printf("  - Item is not returnable.\n");
    }
    $this->printIfTrue($item->getQuantityOrdered(), 'Quantity ordered');
    $this->printIfTrue($item->getQuantityPending(), 'Quantity pending');
    $this->printIfTrue($item->getQuantityCanceled(), 'Quantity canceled');
    $this->printIfTrue($item->getQuantityShipped(), 'Quantity shipped');
    $this->printIfTrue($item->getQuantityDelivered(), 'Quantity delivered');
    $this->printIfTrue($item->getQuantityReturned(), 'Quantity returned');
    if (!empty($item->getCancellations())) {
      printf("  - %d cancellation(s):\n", count($item->getCancellations()));
      foreach ($item->getCancellations() as $cancel) {
        printf("    Cancellation:\n");
        $this->printIfTrue($cancel->getActor(), 'Actor', '    ');
        printf("    - Creation date: %s\n", $cancel->getCreationDate());
        printf("    - Quantity: %s\n", $cancel->getQuantity());
        printf("    - Reason: %s\n", $cancel->getReason());
        printf("    - Reason text: %s\n", $cancel->getReasonText());
      }
    }
    if (!empty($item->getReturns())) {
      printf("  - %d return(s):\n", count($item->getReturns()));
      foreach ($item->getReturns() as $ret) {
        printf("    Return:\n");
        $this->printIfTrue($ret->getActor(), 'Actor', '    ');
        printf("    - Creation date: %s\n", $ret->getCreationDate());
        printf("    - Quantity: %s\n", $ret->getQuantity());
        printf("    - Reason: %s\n", $ret->getReason());
        printf("    - Reason text: %s\n", $ret->getReasonText());
      }
    }
  }
}

$sample = new OrdersSample();
$sample->run();

File: dotnet/source/OrdersSample.cs
using System;
using Google.Apis.ShoppingContent.v2_1;
using Google.Apis.Services;
using Google.Apis.ShoppingContent.v2_1.Data;
using System.Collections.Generic;
using CommandLine;

namespace ShoppingSamples.Content
{
    /// <summary>
    /// A sample consumer that runs an example workflow for a single test order using the Orders
    /// service in the Content API for Shopping.
    ///
    /// <para>Since access to the Orders service is limited, this sample is written as a separate
    /// program from the rest. Also unlike the other samples, this sample uses the sandbox API
    /// endpoint instead of the normal endpoint for two reasons:</para>
    /// <list type="bullet">
    /// <item>
    /// <description>It provides access to the methods for creating and operating on test
    /// orders.</description>
    /// </item>
    /// <item>
    /// <description>It avoids accidentally mutating existing real orders.</description>
    /// </item></list>
    /// </summary>
    internal class OrdersSample : BaseContentSample
    {
        private Random prng; // Used for random order/tracking/shipment ID creation.
        private ulong nonce; // Nonce used for creating new operation IDs.

        public OrdersSample()
        {
            this.prng = new Random();
        }

        internal override void runCalls()
        {
            var merchantId = config.MerchantId.Value;

            // First, we create a new test order using the template1 template.  Normally, orders
            // would be automatically populated by Google in the non-sandbox version, and we'd
            // skip ahead to find out what orders are currently waiting for our acknowledgment.
            Console.WriteLine("=================================================================");
            Console.WriteLine("Creating Test Order for {0}", merchantId);
            Console.WriteLine("=================================================================");

            string orderId;
            {
                var req = new OrdersCreateTestOrderRequest() {
                  TemplateName = "template1"
                };
                var resp = sandboxService.Orders.Createtestorder(req, merchantId).Execute();
                orderId = resp.OrderId;
            }

            Console.WriteLine("Order created with ID {0}", orderId);
            Console.WriteLine();

            // List all unacknowledged orders.  In normal usage, this is where we'd get new
            // order IDs to operate on.  We should see the new test order show up here.
            //ListAllUnacknowledgedOrders(merchantId);

            // Acknowledge the newly created order.  The assumption in doing so is that we've
            // collected the information for this order into our own internal system, and
            // acknowledging it allows us to skip it when searching for new orders (see
            // ListAllUnacknowledgedOrders()).
            Acknowledge(merchantId, orderId);

            // We'll use random numbers for a few things that would normally be supplied by
            // various systems, like the order IDs for the merchant's internal systems, the
            // shipping IDs for the merchant's internal systems, and the tracking ID we'd get
            // from the shipping carriers.
            string merchantOrderId = prng.Next().ToString();
            UpdateMerchantOrderId(merchantId, orderId, merchantOrderId);

            // Print out the current status of the order (and store a reference to the resource
            // so we can pull stuff out of it shortly).
            var currentOrder = GetOrderByMerchantOrderId(merchantId, merchantOrderId);
            PrintOrder(currentOrder);
            Console.WriteLine();

            // Oops, not enough stock for all the Chromecasts ordered, so we cancel one of them.
            {
                var req = new OrdersCancelLineItemRequest() {
                  LineItemId = currentOrder.LineItems[0].Id,
                  Quantity = 1,
                  Reason = "noInventory",
                  ReasonText = "Ran out of inventory while fulfilling request.",
                  OperationId = NewOperationId()
                };

                CancelLineItem(merchantId, orderId, req);
            }

            currentOrder = GetOrder(merchantId, orderId);
            PrintOrder(currentOrder);
            Console.WriteLine();

            // At this point we'll advance the test order, which simulates the point at which
            // an order is no longer cancelable by the customer and the order information reflects
            // that the order is ready for shipping.
            Console.WriteLine("=================================================================");
            Console.WriteLine("Advancing Test Order {0}", orderId);
            Console.WriteLine("=================================================================");
            sandboxService.Orders.Advancetestorder(merchantId, orderId).Execute();
            Console.WriteLine();

            currentOrder = GetOrder(merchantId, orderId);
            PrintOrder(currentOrder);
            Console.WriteLine();

            // To simulate partial fulfillment, we'll pick the first line item and ship the
            // still-pending amount. (Remember, we cancelled one already.)
            var shipItem1Req = ShipAllLineItem(merchantId, orderId, currentOrder.LineItems[0]);

            currentOrder = GetOrder(merchantId, orderId);
            PrintOrder(currentOrder);
            Console.WriteLine();

            // Now we ship the rest.
            var shipItem2Req = ShipAllLineItem(merchantId, orderId, currentOrder.LineItems[1]);

            currentOrder = GetOrder(merchantId, orderId);
            PrintOrder(currentOrder);
            Console.WriteLine();

            // Customer receives the first item, so we notify Google that it was delivered.
            LineItemDelivered(merchantId, orderId, shipItem1Req);

            currentOrder = GetOrder(merchantId, orderId);
            PrintOrder(currentOrder);
            Console.WriteLine();

            // Customer receives second item.
            LineItemDelivered(merchantId, orderId, shipItem2Req);

            currentOrder = GetOrder(merchantId, orderId);
            PrintOrder(currentOrder);
            Console.WriteLine();

            // Turns out one of the first items was broken when the customer received it, so
            // they returned it.  Let's make sure Google knows about it and why.
            {
                var req = new OrdersReturnRefundLineItemRequest();
                req.LineItemId = currentOrder.LineItems[0].Id;
                req.Quantity = 1;
                req.Reason = "productArrivedDamaged";
                req.ReasonText = "Item broken at receipt.";
                req.OperationId = NewOperationId();

                LineItemReturned(merchantId, orderId, req);
            }

            currentOrder = GetOrder(merchantId, orderId);
            PrintOrder(currentOrder);
            Console.WriteLine();
        }

        /// <summary>
        /// Retrieves a particular order given the (Google-specified) order ID.
        /// </summary>
        /// <returns>The order resource for the specified order.</returns>
        private Order GetOrder(ulong merchantId, string orderId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Getting Order {0}", orderId);
            Console.WriteLine("=================================================================");


            Order status = sandboxService.Orders.Get(merchantId, orderId).Execute();
            Console.WriteLine();

            return status;
        }

        /// <summary>
        /// Retrieves a particular order given the merchant-specified order ID.
        /// </summary>
        /// <returns>The order resource for the specified order.</returns>
        private Order GetOrderByMerchantOrderId(ulong merchantId, string merchantOrderId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Getting Merchant Order {0}", merchantOrderId);
            Console.WriteLine("=================================================================");

            var resp =
                sandboxService.Orders.Getbymerchantorderid(merchantId, merchantOrderId).Execute();
            Console.WriteLine();

            return resp.Order;
        }

        /// <summary>
        /// Lists all the unacknowledged orders for the given merchant.
        /// </summary>
        private void ListAllUnacknowledgedOrders(ulong merchantId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Listing Unacknowledged Orders for Merchant {0}", merchantId);
            Console.WriteLine("=================================================================");

            // Retrieve orders list in pages and display data as we receive it.
            string pageToken = null;
            OrdersListResponse ordersResponse = null;

            do
            {
                OrdersResource.ListRequest ordersRequest = sandboxService.Orders.List(merchantId);
                ordersRequest.Acknowledged = false;
                ordersRequest.PageToken = pageToken;

                ordersResponse = ordersRequest.Execute();

                if (ordersResponse.Resources != null && ordersResponse.Resources.Count != 0)
                {
                    foreach (var order in ordersResponse.Resources)
                    {
                        PrintOrder(order);
                    }
                }
                else
                {
                    Console.WriteLine("No orders found.");
                }

                pageToken = ordersResponse.NextPageToken;
            } while (pageToken != null);
            Console.WriteLine();
        }

        // All (non-test) requests that change an order must have a unique operation
        // ID over the lifetime of the order to enable Google to detect and reject
        // duplicate requests.
        // Here, we just use a nonce and bump it each time, since we're not retrying failures
        // and sending them all sequentially.
        private string NewOperationId() {
            string str = nonce.ToString();
            nonce++;
            return str;
        }

        private void Acknowledge(ulong merchantId, string orderId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Acknowledging Order {0}", orderId);
            Console.WriteLine("=================================================================");

            var req = new OrdersAcknowledgeRequest() {
              OperationId = NewOperationId()
            };
            var resp = sandboxService.Orders.Acknowledge(req, merchantId, orderId).Execute();

            Console.WriteLine("Finished with status {0}.", resp.ExecutionStatus);
            Console.WriteLine();
        }

        private void UpdateMerchantOrderId(ulong merchantId, string orderId, string merchantOrderId)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Updating Merchant Order ID to {0}", merchantOrderId);
            Console.WriteLine("=================================================================");

            var req = new OrdersUpdateMerchantOrderIdRequest() {
              OperationId = NewOperationId(),
              MerchantOrderId = merchantOrderId
            };
            var resp = sandboxService.Orders
                .Updatemerchantorderid(req, merchantId, orderId)
                .Execute();

            Console.WriteLine("Finished with status {0}.", resp.ExecutionStatus);
            Console.WriteLine();
        }

        private void CancelLineItem(ulong merchantId, string orderId,
            OrdersCancelLineItemRequest req)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Canceling {0} of item {1}", req.Quantity, req.LineItemId);
            Console.WriteLine("=================================================================");

            var resp = sandboxService.Orders.Cancellineitem(req, merchantId, orderId).Execute();

            Console.WriteLine("Finished with status {0}.", resp.ExecutionStatus);
            Console.WriteLine();
        }

        private OrdersShipLineItemsRequest ShipAllLineItem(ulong merchantId, string orderId,
            OrderLineItem item)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Shipping {0} of item {1}", item.QuantityPending, item.Id);
            Console.WriteLine("=================================================================");

            var itemShip = new OrderShipmentLineItemShipment() {
              LineItemId = item.Id,
              Quantity = item.QuantityPending
            };

            var shipmentInfo = new OrdersCustomBatchRequestEntryShipLineItemsShipmentInfo() {
              Carrier = item.ShippingDetails.Method.Carrier,
              ShipmentId = prng.Next().ToString(),
              TrackingId = prng.Next().ToString(),
            };

            var req = new OrdersShipLineItemsRequest {
              ShipmentInfos = new List<OrdersCustomBatchRequestEntryShipLineItemsShipmentInfo> { shipmentInfo },
              LineItems = new List<OrderShipmentLineItemShipment> { itemShip },
              OperationId = NewOperationId()
            };

            var resp = sandboxService.Orders.Shiplineitems(req, merchantId, orderId).Execute();

            Console.WriteLine("Finished with status {0}.", resp.ExecutionStatus);
            Console.WriteLine();

            // We return req here so that we have access to the randomly-generated IDs in the
            // main program.
            return req;
        }

        private void LineItemDelivered(ulong merchantId, string orderId,
            OrdersShipLineItemsRequest ship)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Delivered {0} of item {1}", ship.LineItems[0].Quantity,
                ship.LineItems[0].LineItemId);
            Console.WriteLine("=================================================================");

            var req = new OrdersUpdateShipmentRequest() {
              Carrier = ship.ShipmentInfos[0].Carrier,
              TrackingId = ship.ShipmentInfos[0].TrackingId,
              ShipmentId = ship.ShipmentInfos[0].ShipmentId,
              Status = "delivered",
              OperationId = NewOperationId()
            };

            var resp = sandboxService.Orders.Updateshipment(req, merchantId, orderId).Execute();

            Console.WriteLine("Finished with status {0}.", resp.ExecutionStatus);
            Console.WriteLine();
        }

        private void LineItemReturned(ulong merchantId, string orderId,
            OrdersReturnRefundLineItemRequest req)
        {
            Console.WriteLine("=================================================================");
            Console.WriteLine("Returned {0} of item {1}", req.Quantity,
                req.LineItemId);
            Console.WriteLine("=================================================================");

            var resp = sandboxService.Orders.Returnrefundlineitem(req, merchantId, orderId).Execute();

            Console.WriteLine("Finished with status {0}.", resp.ExecutionStatus);
            Console.WriteLine();
        }

        private void PrintOrder(Order order)
        {
            Console.WriteLine("Order {0}:", order.Id);
            Console.WriteLine("- Status: {0}", order.Status);
            Console.WriteLine("- Merchant: {0}", order.MerchantId);
            Console.WriteLine("- Merchant order ID: {0}", order.MerchantOrderId);
            if (order.Customer != null)
            {
                Console.WriteLine("- Customer information:");
                Console.WriteLine("  - Full name: {0}", order.Customer.FullName);
                if (order.Customer.MarketingRightsInfo != null)
                {
                    Console.WriteLine("  - Email: {0}", order.Customer.MarketingRightsInfo.MarketingEmailAddress);
                }
            }
            Console.WriteLine("- Placed on date: {0}", order.PlacedDate);
            if (order.NetPriceAmount != null)
            {
                Console.WriteLine("- Net amount: {0} {1}", order.NetPriceAmount.Value,
                    order.NetPriceAmount.Currency);
            }
            Console.WriteLine("- Payment status: {0}", order.PaymentStatus);
            Console.WriteLine("- Acknowledged: {0}", order.Acknowledged == true ? "yes" : "no");
            if (order.LineItems != null && order.LineItems.Count > 0)
            {
                Console.WriteLine("- {0} line item(s):", order.LineItems.Count);
                foreach (var item in order.LineItems)
                {
                    PrintOrderLineItem(item);
                }
            }
            if (order.ShippingCost != null)
            {
                Console.WriteLine("- Shipping cost: {0} {1}", order.ShippingCost.Value,
                    order.ShippingCost.Currency);
            }
            if (order.ShippingCostTax != null)
            {
                Console.WriteLine("- Shipping cost tax: {0} {1}", order.ShippingCostTax.Value,
                    order.ShippingCostTax.Currency);
            }
            if (order.Shipments != null && order.Shipments.Count > 0)
            {
                Console.WriteLine("- {0} shipment(s):", order.Shipments.Count);
                foreach (var shipment in order.Shipments)
                {
                    Console.WriteLine("  Shipment {0}", shipment.Id);
                    Console.WriteLine("  - Creation date: {0}", shipment.CreationDate);
                    Console.WriteLine("  - Carrier: {0}", shipment.Carrier);
                    Console.WriteLine("  - Tracking ID: {0}", shipment.TrackingId);
                    if (shipment.LineItems != null && shipment.LineItems.Count > 0)
                    {
                        Console.WriteLine("- {0} line item(s):", shipment.LineItems.Count);
                        foreach (var item in shipment.LineItems) {
                            Console.WriteLine("  {0} of item {1}", item.Quantity, item.LineItemId);
                        }
                    }
                    if (shipment.DeliveryDate != null) {
                        Console.WriteLine("  - Delivery date: {0}", shipment.DeliveryDate);
                    }
                }
            }
        }

        private void PrintIfNonzero(long? l, string s)
        {
            if(l.HasValue && l.Value > 0) {
                Console.WriteLine("  - {0}: {1}", s, l.Value);
            }
        }

        private void PrintOrderLineItem(OrderLineItem item)
        {
            Console.WriteLine("  Line item: {0}", item.Id);
            Console.WriteLine("  - Product: {0} ({1})", item.Product.Id, item.Product.Title);
            Console.WriteLine("  - Price: {0} {1}", item.Price.Value, item.Price.Currency);
            Console.WriteLine("  - Tax: {0} {1}", item.Tax.Value, item.Tax.Currency);
            if (item.ShippingDetails != null)
            {
                Console.WriteLine("  - Ship by date: {0}", item.ShippingDetails.ShipByDate);
                Console.WriteLine("  - Deliver by date: {0}", item.ShippingDetails.DeliverByDate);
                Console.WriteLine("  - Deliver via {0} {1} ({2} - {3} days)",
                    item.ShippingDetails.Method.Carrier, item.ShippingDetails.Method.MethodName,
                    item.ShippingDetails.Method.MinDaysInTransit,
                    item.ShippingDetails.Method.MaxDaysInTransit);
            }
            if (item.ReturnInfo != null && item.ReturnInfo.IsReturnable == true)
            {
                Console.WriteLine("  - Item is returnable.");
                Console.WriteLine("    - Days to return: {0}", item.ReturnInfo.DaysToReturn);
                Console.WriteLine("    - Return policy is at {0}.", item.ReturnInfo.PolicyUrl);
            }
            else
            {
                Console.WriteLine("  - Item is not returnable.");
            }
            PrintIfNonzero(item.QuantityOrdered, "Quantity Ordered");
            PrintIfNonzero(item.QuantityPending, "Quantity Pending");
            PrintIfNonzero(item.QuantityCanceled, "Quantity Canceled");
            PrintIfNonzero(item.QuantityShipped, "Quantity Shipped");
            PrintIfNonzero(item.QuantityDelivered, "Quantity Delivered");
            PrintIfNonzero(item.QuantityReturned, "Quantity Returned");
            if (item.Cancellations != null && item.Cancellations.Count > 0)
            {
                Console.WriteLine("  - {0} cancellations(s):", item.Cancellations.Count);
                foreach (var cancel in item.Cancellations)
                {
                    Console.WriteLine("    Cancellation:");
                    if (cancel.Actor != null)
                    {
                        Console.WriteLine("    - Actor: {0}", cancel.Actor);
                    }
                    Console.WriteLine("    - Creation date: {0}", cancel.CreationDate);
                    Console.WriteLine("    - Quantity: {0}", cancel.Quantity);
                    Console.WriteLine("    - Reason: {0}", cancel.Reason);
                    Console.WriteLine("    - Reason text: {0}", cancel.ReasonText);
                }
            }
            if (item.Returns != null && item.Returns.Count > 0)
            {
                Console.WriteLine("  - {0} return(s):", item.Returns.Count);
                foreach (var ret in item.Returns)
                {
                    Console.WriteLine("    Return:");
                    if (ret.Actor != null)
                    {
                        Console.WriteLine("    - Actor: {0}", ret.Actor);
                    }
                    Console.WriteLine("    - Creation date: {0}", ret.CreationDate);
                    Console.WriteLine("    - Quantity: {0}", ret.Quantity);
                    Console.WriteLine("    - Reason: {0}", ret.Reason);
                    Console.WriteLine("    - Reason text: {0}", ret.ReasonText);
                }
            }
        }
    }
}

